File: LRU.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type Node struct {
	key  string
	data string
	prev *Node
	next *Node
}

type LRUCache struct {
	capacity int
	head     *Node
	tail     *Node
}

func NewLRUCache(cap int) *LRUCache {
	head := &Node{key: "", data: ""}
	tail := &Node{key: "", data: ""}
	head.next = tail
	tail.prev = head
	return &LRUCache{
		capacity: cap,
		head:     head,
		tail:     tail,
	}
}

func (lru *LRUCache) findNode(key string) *Node {
	current := lru.head.next
	for current != lru.tail {
		if current.key == key {
			return current
		}
		current = current.next
	}
	return nil
}

func (lru *LRUCache) removeNode(node *Node) {
	node.prev.next = node.next
	node.next.prev = node.prev
}

func (lru *LRUCache) addToFront(node *Node) {
	node.next = lru.head.next
	node.prev = lru.head
	lru.head.next.prev = node
	lru.head.next = node
}

func (lru *LRUCache) moveToFront(node *Node) {
	lru.removeNode(node)
	lru.addToFront(node)
}

func (lru *LRUCache) removeLRU() {
	lruNode := lru.tail.prev
	lru.removeNode(lruNode)
}

func (lru *LRUCache) Get(key string) *Node {
	node := lru.findNode(key)
	if node != nil {
		lru.moveToFront(node)
		return node
	}
	return nil
}

func (lru *LRUCache) Set(key, data string) {
	node := lru.findNode(key)
	if node != nil {
		lru.moveToFront(node)
		node.data = data
	} else {
		if lru.GetSize() >= lru.capacity {
			lru.removeLRU()
		}
		newNode := &Node{key: key, data: data}
		lru.addToFront(newNode)
	}
}

func (lru *LRUCache) GetSize() int {
	size := 0
	current := lru.head.next
	for current != lru.tail {
		size++
		current = current.next
	}
	return size
}

func (lru *LRUCache) PrintCache() {
	fmt.Print("Кэш: ")
	current := lru.head.next
	for current != lru.tail {
		fmt.Printf("[\"%s %s\"] ", current.key, current.data)
		current = current.next
	}
	fmt.Println()
}

func LRU() {
	fmt.Println("Режим LRU-кеша")
	fmt.Println("Введите размер кеша:")

	var cacheSize int
	fmt.Scan(&cacheSize)

	if cacheSize <= 0 {
		fmt.Println("Ошибка: размер кеша должен быть положительным числом")
		return
	}

	cache := NewLRUCache(cacheSize)
	fmt.Printf("LRU-кеш создан с размером %d\n", cacheSize)

	scanner := bufio.NewScanner(os.Stdin)

	for {
		fmt.Println("\nДоступные команды:")
		fmt.Println("SET <ключ> - добавить элемент в кеш")
		fmt.Println("GET <ключ> - проверить наличие элемента в кеше")
		fmt.Println("PRINT - вывести текущее состояние кеша")
		fmt.Println("EXIT - выход из программы")
		fmt.Println("Введите команду:")

		scanner.Scan()
		input := scanner.Text()

		parts := strings.Fields(input)
		if len(parts) == 0 {
			continue
		}

		command := parts[0]

		if command == "EXIT" {
			fmt.Println("Выход из программы")
			break
		} else if command == "SET" {
			if len(parts) >= 3 {
				key := parts[1]
				data := strings.Join(parts[2:], " ")
				cache.Set(key, data)
				fmt.Printf("Элемент добавлен в кеш: ключ='%s'\n", key)
				cache.PrintCache()
			} else {
				fmt.Println("Ошибка: неверный формат команды. Используйте: SET <ключ>")
			}
		} else if command == "GET" {
			if len(parts) >= 2 {
				key := parts[1]
				found := cache.Get(key)
				if found != nil {
					fmt.Printf("Элемент найден в кеше: ='%s'\n", found.data)
				} else {
					fmt.Printf("Элемент отсутствует в кеше: ключ='%s'\n", key)
				}
				cache.PrintCache()
			} else {
				fmt.Println("Ошибка: неверный формат команды. Используйте: GET <ключ>")
			}
		} else if command == "PRINT" {
			cache.PrintCache()
		} else {
			fmt.Printf("Ошибка: неизвестная команда '%s'\n", command)
		}
	}
}

File: LRU.cpp
#include <iostream>
#include <string>
#include <sstream>

using namespace std;

struct Node {
    string key;
    string data;
    Node* prev;
    Node* next;
    
    Node(string k, string d) : key(k), data(d), prev(nullptr), next(nullptr) {}
};

class LRUCache {
private:
    int capacity;
    Node* head;
    Node* tail;

    Node* findNode(string key) {
        Node* current = head->next;
        while (current != tail) {
            if (current->key == key) {
                return current;
            }
            current = current->next;
        }
        return nullptr;
    }

    void removeNode(Node* node) {
        node->prev->next = node->next;
        node->next->prev = node->prev;
    }

    void addToFront(Node* node) {
        node->next = head->next;
        node->prev = head;
        head->next->prev = node;
        head->next = node;
    }

    void moveToFront(Node* node) {
        removeNode(node);
        addToFront(node);
    }

    void removeLRU() {
        Node* lru = tail->prev;
        removeNode(lru);
        delete lru;
    }

public:
    LRUCache(int cap) : capacity(cap) {
        head = new Node("", "");
        tail = new Node("", "");
        head->next = tail;
        tail->prev = head;
    }

    ~LRUCache() {
        Node* current = head;
        while (current) {
            Node* next = current->next;
            delete current;
            current = next;
        }
    }

    Node* get(string key) {
        Node* node = findNode(key);
        if (node != nullptr) {
            moveToFront(node);
            return node;
        }
        return nullptr;
    }

    void set(string key, string data) {
        Node* node = findNode(key);
        if (node != nullptr) {
            // Ключ уже существует, перемещаем в начало
            moveToFront(node);
            node -> data = data;
        } else {
            if (getSize() >= capacity) {
                removeLRU();
            }
            Node* newNode = new Node(key, data);
            addToFront(newNode);
        }
    }

    int getSize() {
        int size = 0;
        Node* current = head->next;
        while (current != tail) {
            size++;
            current = current->next;
        }
        return size;
    }

    void printCache() {
        cout << "Кэш: ";
        Node* current = head->next;
        while (current != tail) {
            cout << "[\"" << current->key << " " << current -> data<< "\"] ";
            current = current->next;
        }
        cout << endl;
    }
};

void LRU() {
    cout << "Режим LRU-кеша" << endl;
    cout << "Введите размер кеша:" << endl;
    
    int cacheSize;
    cin >> cacheSize;
    
    if (cacheSize <= 0) {
        cout << "Ошибка: размер кеша должен быть положительным числом" << endl;
        return;
    }
    
    LRUCache cache(cacheSize);
    cout << "LRU-кеш создан с размером " << cacheSize << endl;
    
    cin.ignore();
    
    while (true) {
        cout << endl << "Доступные команды:" << endl;
        cout << "SET <ключ> - добавить элемент в кеш" << endl;
        cout << "GET <ключ> - проверить наличие элемента в кеше" << endl;
        cout << "PRINT - вывести текущее состояние кеша" << endl;
        cout << "EXIT - выход из программы" << endl;
        cout << "Введите команду:" << endl;
        
        string input;
        getline(cin, input);
        
        stringstream ss(input);
        string command;
        ss >> command;
        
        if (command == "EXIT") {
            cout << "Выход из программы" << endl;
            break;
        }
        else if (command == "SET") {
            string key;
            string data;
            if (ss >> key && ss >> data) {
                cache.set(key, data);
                cout << "Элемент добавлен в кеш: ключ='" << key << "'" << endl;
                cache.printCache();
            } else {
                cout << "Ошибка: неверный формат команды. Используйте: SET <ключ>" << endl;
            }
        }
        else if (command == "GET") {
            string key;
            string data;
            if (ss >> key) {
                Node* found = cache.get(key);
                if (found) {
                    cout << "Элемент найден в кеше: ='" << found -> data << "'" << endl;
                } else {
                    cout << "Элемент отсутствует в кеше: ключ='" << key << "'" << endl;
                }
                cache.printCache();
            } else {
                cout << "Ошибка: неверный формат команды. Используйте: GET <ключ>" << endl;
            }
        }
        else if (command == "PRINT") {
            cache.printCache();
        }
        else {
            cout << "Ошибка: неизвестная команда '" << command << "'" << endl;
        }
    }
}